sound/soc/sof/intel/ext_manifest.h

```diff
@@ -0,0 +1,33 @@
+/* SPDX-License-Identifier: ((GPL-2.0 WITH Linux-syscall-note) OR BSD-3-Clause) */
+/*
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
+ */
+
+/*
+ * Intel extended manifest is a extra place to store Intel cavs specific
+ * metadata about firmware, for example LPRO/HPRO configuration is
+ * Intel cavs specific. This part of output binary is not signed.
+ */
+
+#ifndef __INTEL_CAVS_EXT_MANIFEST_H__
+#define __INTEL_CAVS_EXT_MANIFEST_H__
+
+#include <sound/sof/ext_manifest.h>
+
+/* EXT_MAN_ELEM_PLATFORM_CONFIG_DATA elements identificators */
+enum sof_cavs_config_elem_type {
+	SOF_EXT_MAN_CAVS_CONFIG_CAVS_LPRO	= 1,
+	SOF_EXT_MAN_CAVS_CONFIG_LAST_ELEM	= SOF_EXT_MAN_CAVS_CONFIG_CAVS_LPRO,
+};
+
+/* EXT_MAN_ELEM_PLATFORM_CONFIG_DATA elements */
+struct sof_ext_man_cavs_config_data {
+	struct sof_ext_man_elem_header hdr;
+
+	struct sof_config_elem elems[SOF_EXT_MAN_CAVS_CONFIG_LAST_ELEM];
+} __packed;
```

Comment:
this got me thinking. This means, that every time you add a new config element type this structure will change size. So if you run a kernel and a firmware with different minor versions the kernel won't know what size of this structure should actually be there. I think you need a normal variable length array here like
```
struct sof_ext_man_cavs_config_data {
	struct sof_ext_man_elem_header hdr;
	uint32_t n_elem;
	struct sof_config_elem elems[];
} __packed;
```

Reply:
It was my point to, but variable explicit defining number of elements will be hard to achieve from FW side, there should be calculation based on `sof_ext_man_cavs_config_data.hdr.elem_size`.

Reply:
why not have a MAX_ELEMS instead then?

Reply:
Initially I thought that defining MAX enum is easy and better way but I'm investigating @lyakh @ktrzcinx 's comments right now.
@ranj063 After returning error, it failed all HDA platform. firmware had extra configs for OUTBOX/INBOX size which we don't use. That break this logic. I can see possible errror when firmware add more field. We should not return error, I will take care of this also.

Reply:
> It was my point to, but variable explicit defining number of elements will be hard to achieve from FW side, there should be calculation based on `sof_ext_man_cavs_config_data.hdr.elem_size`.

@ktrzcinx you mean the xcc's problem with variable size arrays?

Reply:
xcc's problem with variable size arrays is one issue, another is is explicit defining number of elements from FW side.  From FW side those are global, const structs, filling them looks like this:

```
const struct ext_man_config_data ext_man_config
	__aligned(EXT_MAN_ALIGN) __section(".fw_metadata") = {
	.hdr.type = EXT_MAN_ELEM_CONFIG_DATA,
	.hdr.elem_size = ALIGN_UP(sizeof(struct ext_man_config_data),
				  EXT_MAN_ALIGN),
	.elems = {
		{EXT_MAN_CONFIG_IPC_MSG_SIZE, SOF_IPC_MSG_MAX_SIZE},
	},
};
```

As I know inserting `.elems` elements counter to this list would need some hacks. Or are we going about inserting `.elems` capacity, then it shouldn't be any problem (except for ABI change), but result will be the same as it is now (elem counter calculation from struct size).

Reply:
I dropped variable size arrays. After investigating SOF firmware side change, first XCC bug and second, implementation is unnecessary complicated(even ignoring XCC bug).

Instead I bring back LAST_ELEM but LAST_ELEM is initialized with last element of that enum. And also added first enum EMPTY as Explicit 0, this is more clear(even i forgot why it started from 1).

Even with LAST_ELEM, calculating number of elem from header size is good idea. So I kept that also.

To sync up with SOF firmware header, I submit a PR to SOF side as well. https://github.com/thesofproject/sof/pull/3510

Reply:
I still disagree, that setting the array size to the number of currently defined parameters is a good idea. If these parameters are going to be used, then we'll get more than 1 of them pretty quickly, with each time having to modify the ABI. If this will never be extended beyond the original 1 value, well, maybe we don't need an array then in the first place at all.

Reply:
>then we'll get more than 1 of them pretty quickly, with each time having to modify the ABI

If this array is flex array from kernel point of view and unknown elements are ignored, then ABI change won't be needed after extend (eg. multiple usage of single key).

Reply:
Some finding about the elem_size here. see https://github.com/thesofproject/linux/pull/2459/files#r508172578
@fredoh9 @ktrzcinx how could we get real elem size with align up here?
```
	.hdr.elem_size = ALIGN_UP(sizeof(struct ext_man_config_data),
				  EXT_MAN_ALIGN),
```